keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65536, mapping the range 0-65536 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65536 (2^16). This imposes a limit of 65536 pods per node, but
+that is not an issue we will hit in practice for a long time, if ever (today we
+run 110 pods per node by default).
+
+#### Phase 2: pods with volumes
+
+This phase makes user namespaces work in pods with volumes too. This is
+activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
+some other volume type than the ones listed for phase 1 is used. IOW, when phase
+2 is implemented, pods that use volume types not supported in phase 1, fall into
+the phase 2.
+
+All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
+length 65536, and mapping the range 0-65536 too. IOW, each pod will have its own user
+namespace, but they will map to _the same_ UIDs/GIDs in the host.
+
+Using the same mapping allows for pods to share files and mitigates all the
+listed vulnerabilities (as the host is protected from the container). It is also
+a trivial next-step to take, given that we have phase 1 implemented: just return
+the same mapping if the pod has other volumes.
+
+#### Phase 3: pod to pod isolation
+
+This phase will provide more isolation between pods that use volumes (as in
+phase 2) and requires another opt-in field:
+`pod.spec.securityContext.pod2podIsolation`.
+
+This phase will try to not share the same mapping for all pods with volumes, as
+phase 2 does, but to achieve it some trade off needs to be made. This phase
+builds on the work of the previous phases and more details will be defined while
+the other phases evolve.
+
+Here are some ideas so far:
+
+One idea is to give different mappings to pods in different k8s namespaces or
+that use a different service account. This needs to be explored in further
+detail, but will probably impose limits to which workloads can run this (we need
+to expose a shorter mapping, less than 65536).
+
+Another idea is to use id mapped mounts. This probably needs changes to the
```

Comment:
Are idmapped mounts the future/ideal solution? If so, do the fields and semantics in this KEP continue to make sense when that is available?

Reply:
Good question. Well, if they are the ideal/future solution depends on who you ask.

id mapped mounts provide support to run pods with non-overlapping mappings while sharing volumes between them. So, this is _really_ nice. The downside is that the filesystem the volumes are using need to support id mapped mounts, so you need a recent kernel that added support for the fs that you want to use.

It is clear that we will have to live several years (5-10?) until users get a kernel with id mapped support for several filesystems (the supported fs are not a lot today). But is not that _all_ the filesystems will support it in that time frame, at least we don't have any reason to believe that now. So, the only reasonable thing to plan is that we will need to support userns without id mapped mounts for a very long time (I'd say forever).

And yes, this KEP semantics make sense even when id mapped mounts are widely available: we have a pod.spec field that when enabled (or if the "auto" option is implemented as it is mentioned in these lines) the user can select it and have the best isolation possible.

Furthermore, we also allow room to grow in the future if needed too, given that is an enum. As it came up in the discussion of PR #2101, there are different strategies people want to use that have different trade-offs. All of them are allowed in the current design, as you can choose different strategies for this field.

@cbandy let me know what you think

Reply:
Marking as resolved given your reaction, then :)